File: 690-EmployeeImportance/690-EmployeeImportance.py
# Last updated: 4/29/2026, 10:28:47 AM
"""
# Definition for Employee.
class Employee:
    def __init__(self, id: int, importance: int, subordinates: List[int]):
        self.id = id
        self.importance = importance
        self.subordinates = subordinates
"""
"""
20:00s
O(E) (E: # of employees)
O(E)
"""
class Solution:
    def getImportance(self, employees: List['Employee'], id: int) -> int:    
        def addImportance(employee, hash):
            # print(employee.id)
            # print(employee.importance)
            if not employee:
                return 0
            sum = employee.importance
            if employee.subordinates:   
                for sub_id in employee.subordinates:
                    sub_employee = hash[sub_id]
                    if sub_employee:
                        sum += addImportance(sub_employee, hash)
            return sum
        from collections import defaultdict
        hash = defaultdict(lambda: None)
        for employee in employees:
            hash[employee.id] = employee
            
        e = hash[id]
        return addImportance(e, hash)

File: 690-EmployeeImportance/690-EmployeeImportance.js
// Last updated: 4/29/2026, 10:28:42 AM
/**
 * Definition for Employee.
 * function Employee(id, importance, subordinates) {
 *     this.id = id;
 *     this.importance = importance;
 *     this.subordinates = subordinates;
 * }
 */

/**
 * @param {Employee[]} employees
 * @param {number} id
 * @return {number}
 */
var GetImportance = function(employees, id) {
    const map = new Map();
    for(const employee of employees) {
        map.set(employee.id, employee);
    }
    const addImportance = (id, map) => {
        const e = map.get(id);
        let sum = 0;
        if(e) {
            sum += e.importance;
            const subIds = e.subordinates;
            if(subIds) {
                for(const subId of subIds) {
                    sum += addImportance(subId, map)
                }
            }
        } else {
            return 0;
        }
        return sum;
    }
    
    return addImportance(id, map);
};